src/bin/cargo/commands/remove.rs

```diff
@@ -114,3 +122,76 @@ fn parse_section(args: &ArgMatches) -> DepTable {
 
     table
 }
+
+/// Clean up workspace dependencies which no longer have a reference to them.
+fn gc_workspace(workspace: &Workspace<'_>, dependencies: &[String]) -> CargoResult<()> {
+    let mut manifest: toml_edit::Document =
+        cargo_util::paths::read(workspace.root_manifest())?.parse()?;
+
+    let members = workspace
+        .members()
+        .map(|p| LocalManifest::try_new(p.manifest_path()).map(|m| (p, m)))
+        .collect::<CargoResult<Vec<_>>>()?;
+
+    for dep in dependencies {
+        if !dep_in_workspace(dep, &members) {
+            remove_workspace_dep(dep, &mut manifest);
+        }
+    }
+
+    cargo_util::paths::write(workspace.root_manifest(), manifest.to_string().as_bytes())?;
+
+    Ok(())
+}
+
+/// Get whether or not a dependency is marked as `workspace`.
+fn dep_is_workspace(dep: &str, dep_table: &[String], manifest: &LocalManifest) -> bool {
+    if let Ok(toml_edit::Item::Table(table)) = manifest.get_table(dep_table) {
+        let value = table.get(dep).and_then(|i| i.get("workspace"));
+        if let Some(toml_edit::Item::Value(value)) = value {
+            return value.as_bool() == Some(true);
+        }
+    }
+
+    false
+}
+
+/// Get whether or not a dependency is depended upon in a workspace.
+fn dep_in_workspace(dep: &str, members: &[(&Package, LocalManifest)]) -> bool {
+    members.iter().any(|(pkg, manifest)| {
+        pkg.dependencies()
+            .iter()
+            .filter(|d| d.name_in_toml() == dep)
+            .map(|d| (d, dep_to_table(d)))
+            .any(|(d, t)| {
```

Comment:
nit: for where the dependency was removed, we are relying on the undocumented behavior for when `dep_is_workspace`  does not find the promised dependency.  Ideally, we would make this explicit.  This could be done either by
- Having `dep_is_workspace` return `Option<bool>`
- Iterate over the dependencies in `manifest` rather than `pkg`.  `LocalManifest` should have fairly friendly accessors to help with that

Personally, I would prefer that both happen.  If both are done, we could do a `.expect()` on `dep_is_workspace` or turn it into an internal error.

Reply:
Doing the first should be easy. It's important to note that `dep_is_workspace` will return `None` on the recently removed dependency (as it hasn't been updated in the `Workspace` but has in the manifest), so doing `unwrap_or(false)` would make more sense. That's essentially what is going on right now, although poorly documented.

We could inspect each manifest with `LocalManifest` rather than `Package`, but that would require adding support for manually searching through `dependencies`, `dev-dependencies`, `target.xyz.build-dependencies`, etc. Using `Package`/`Dependency` was an intentional choice to avoid this. Do you still think it is worth the rewrite?

Reply:
> hat would require adding support for manually searching through dependencies, dev-dependencies, target.xyz.build-dependencies, etc.

`Manifest::get_sections` will let us iterate over all dependency sections and then we just need to check if the returned table-likes contain the dependency key

Reply:
Done.